README.md
# nodegame-benchmark (beta)

Benchmark and Testing for nodeGame.

## Overview

The benchmark script reads settings from a config file, starts the
nodeGame server, connects a number of PhantomJS clients to the specified
channel and lets them play.

While playing, the script collects statistics about the number and the
timing of all exchanged messages, and also about CPU and memory usage
(if the dependency `psutil` is installed).

When all games are finished, all statistics are saved to csv files, as
specified in the settings. If an automated test script is defined by the
`package.json` file inside the game directory, such a test-script will
executed and the results displayed on screen.


## Usage

1. Create a copy of the file `config-sample.ini` with name `config.ini`
2. Edit `config.ini` with the correct settings for your environment (see below)
3. Run the benchmark either with `python run_benchmark.py` or simply with
   `./run_benchamrk.py`

### config.ini

This file needs to define the following variables. Note that it
is possible to refer to variables defined in other sections via
`${SECTION_NAME:VAR_NAME}` syntax.

```ini
[General Settings]
    game: ultimatum

; This is the directories section.
; All of these directories need to exist.
[Directories]
    nodegame_dir: ~/nodegame
    benchmark_dir: ${nodegame_dir}/node_modules/nodegame-benchmark
    client_dir: ${nodegame_dir}/node_modules/nodegame-client
    server_dir: ${nodegame_dir}/node_modules/nodegame-server
    game_dir: ${nodegame_dir}/games/${General Settings:game}

    ; This is the directory where stdout and stderr logs will be written.
    log_dir: ${benchmark_dir}/log

    ; This is the directory where data in csv format will be written.
    csv_dir: ${benchmark_dir}/csv

    ; This folder contains the launcher of the benchmark.
    launcher_dir: ${nodegame_dir}/test

    ; This folder determines the "current working directory" of the launcher.
    launcher_cwd: ${nodegame_dir}

    ; This folder contains the logs of messages.
    msg_log_dir: ${nodegame_dir}/log

    ; This folder is the current working directory of the test that sanitizes
    ; benchmark results.
    test_cwd: ${game_dir}

; This is the files section.
; All of these files need to exist.
[Files]
    ; This is where the values of the [Client Variables] section are written.
    client_var_file: ${Directories:client_dir}/lib/modules/variables.js

    ; This is where the values of the [Server Variables] section are written.
    server_var_file: ${Directories:server_dir}/conf/servernode.js

    ; This file contains the messages received and sent by the server.
    server_msg_file: ${Directories:msg_log_dir}/messages.log

    ; These files specify the launcher and its settings file.
    launcher_file: ${Directories:launcher_dir}/launcher-autoplay.js
    launcher_settings_file: ${Directories:game_dir}/test/settings.js

[Client Variables]
    rel_msg_var: k.reliableMessaging
    rel_retry_var: k.reliableRetryInterval

[Server Variables]
    rel_msg_var: servernode.reliableMessaging
    rel_retry_var: servernode.reliableRetryInterval

; The settings specified here will be written to the launcher_settings_file
[Launcher Settings]
    ; sioTransports are not working in Express 4 at the moment.
    ; sioTransports: ["websocket", "flashsocket", "htmlfile", "xhr-polling",
    ;                 "jsonp-polling"]
```

## Options for run_benchmark

```
$ ./run_benchmark.py --help
usage: run_benchmark.py [-h] -c CONFIG [-n NUM_CONNS [NUM_CONNS ...]] [-r]
                        [-nr] [-t TIMEOUTS [TIMEOUTS ...]]

Execute nodegame benchmark and write benchmark data to csv file.

optional arguments:
  -h, --help            show this help message and exit
  -c CONFIG, --config CONFIG
                        Benchmark configuration file in INI format containing
                        variables that likely do not change between
                        benchmarks.
  -n NUM_CONNS [NUM_CONNS ...], --num_conns NUM_CONNS [NUM_CONNS ...]
                        Number of simultaneous connections to consider for the
                        benchmark, can be a list.
  -r, --reliable        Boolean flag to turn on reliable messaging.
  -nr, --no_run         Boolean flag to disable launching the game. Will just
                        process existing log files.
  -t TIMEOUTS [TIMEOUTS ...], --timeouts TIMEOUTS [TIMEOUTS ...]
                        Timeouts to consider for the benchmark when reliable
                        messaging is used, can be a list.
```

## File format of metrics.csv

`metrics.csv` defines the following data headers:

- `id`: Unix timestamp of the experiment in milliseconds. It is very unlikely
  that two different runs have the same timestamp, hence the use as an
  identifier is justified.
- `machine`: Name of the current machine as reported by Python's
  [`platform.platform()`](https://docs.python.org/3.4/library/platform.html#platform.platform).
- `num_conns`: Number of simultaneous connections during this run of the
  experiment.  This is the number specified by the `-n` flag.
- `is_reliable`: Reports whether reliable messaging was turned on or off. This
  is specified by the presence or absence of the `-r` flag.
- `timeout`: Reports the timeout used during the current run. Only meaningful
  if reliable messaging was turned on. Controlled via the `-t` flag.
- `benchmark_ret_code`: Return code of the benchmark process. A number
  different from 0 indicates that there was a problem.
- `test_ret_code`: Return code of the test process. A number different from 0
  indicates that there was a problem.
- `cpu_time_user`: User CPU time as reported by
  [`psutil.Process.cpu_times()`](https://pythonhosted.org/psutil/#psutil.Process.cpu_times).
  This time also includes CPU time spent by children processes of the current
  process.
- `cpu_time_system`: System CPU time as reported by
  [`psutil.Process.cpu_times()`](https://pythonhosted.org/psutil/#psutil.Process.cpu_times).
  This time also includes CPU time spent by children processes of the current
  process.
- `mem_info_rss`: Resident Set Size (RSS) usage as reported by
  [`psutil.Process.memory_info()`](https://pythonhosted.org/psutil/#psutil.Process.memory_info).
  This usage also includes memory usage by children processes of the current
  process.
- `mem_info_vms`: Virtual Memory Size (VMS) usage as reported by
  [`psutil.Process.memory_info()`](https://pythonhosted.org/psutil/#psutil.Process.memory_info).
  This usage also includes memory usage by children processes of the current
  process.
- `avg_client_time`: When reliable messaging is enabled, this is the average
  time to process a client message. Time is measured as the duration between
  receiving a client message and sending the corresponding ACK message.
  Currently this is only measured on the server and hence is not meaningful.
- `avg_server_time`: When reliable messaging is enabled, this is the average
  time for a round trip for messages from the server. The duration reported is
  the difference between sending a message from a server and receiving the ACK
  for it.


## Example Runs

Reads the config file and run 1 benchmark where there is 1 game:

    ./run_benchmark.py -c config.json -n 1


Here we consider 1, 2, 4 and 8 simultaneous connections, reliable
messaging to be activated, and timeouts of 1000, 2000 and 4000
milliseconds.

    ./run_benchmark.py -c config.json -n 1 2 4 8 -r -t 1000 2000 4000


## Requirements and Dependencies

To be able to report CPU and memory usage, `run_benchmark.py` relies
on the third party module `psutil`. You should be able to install it
via:

     pip install psutil

See
[this manual](https://github.com/giampaolo/psutil/blob/master/INSTALL.rst)
for further help.

Furthermore the script relies on features on included with Python 2.7,
so please make sure to have a recent Python installation available.

## License

[MIT](LICENSE)


CHANGELOG
# CHANGELOG

## 0.1
- Changelog started.


pairtests/pairstest.js
// Usage: phantomjs pairstest.js [num_connections [url]]

var webpage = require('webpage'),
    system = require('system');

function logObject(obj) {
    for (var key in obj) {
        if (obj.hasOwnProperty(key)) {
            console.log('*', key, ':', obj[key]);
        }
    }
}

var url = 'http://localhost:8080/pairs/';
var n = 2;
var i, page;
var countClosed = 0;

if (system.args.length >= 2) {
    n = +system.args[1];
}
if (system.args.length >= 3) {
    url = system.args[2];
}

// Disable cookies
phantom.cookiesEnabled = false;

console.log('Opening ' + n + ' connections to "' + url + '"...');

for (i = 1; i <= n; i ++) {
    page = webpage.create();
    (function(pg, pgNum) {
        var timeoutId;

        pg.onConsoleMessage = function(msg) {
            if (msg === 'Game over') {
                // Game Over
                console.log('Page ' + pgNum + ' finished.');

                // Don't capture the screen of the finished client
                clearTimeout(timeoutId);

                countClosed ++;
                if (countClosed >= n) {
                    console.log('All pages finished. Exiting...');
                    phantom.exit();
                }
                else {
                    console.log('Still waiting for ' + (n - countClosed) + ' clients to finish.');
                }
            }

            if (msg.toLowerCase().indexOf('error') >= 0) {
                // Error
                console.log('Error message: ' + msg);
            }
        };

        pg.onError = function(msg, trace) {
            var msgStack = ['ERROR: ' + msg];
            if (trace && trace.length) {
                msgStack.push('TRACE:');
                trace.forEach(function(t) {
                    msgStack.push(' -> ' + t.file + ': ' + t.line + (t.function ? ' (in function "' + t.function + '")' : ''));
                });
            }
            console.error(msgStack.join('\n'));    var msgStack = ['ERROR: ' + msg];
        };

        pg.open(url, function() {
            console.log('Opened page ' + pgNum + '.');
        });

        timeoutId = setTimeout(function() {
            console.log();
            console.log('Capturing page ' + pgNum + '.');
            pg.render('screenshot_' + pgNum + '.png');

            stateObj = pg.evaluate(function() {
                return {
                    player: node.player,
                    pl:     node.game.pl
                };
            });

            console.log('player: ');
            logObject(stateObj.player);
            console.log('player.stage: ');
            logObject(stateObj.player.stage);
            console.log('pl.db: ');
            logObject(stateObj.pl.db);
            console.log();
        }, 240 * 1000);
    })(page, i);
}


pairs-benchmark/phantom-pairs.js
// Usage: phantomjs phantom-pairs.js [num_connections [url]]

var webpage = require('webpage'),
    system = require('system');

var url = 'http://localhost:8080/pairs/';
var n = 2;
var i, page;
var countClosed = 0;

if (system.args.length >= 2) {
    n = +system.args[1];
}
if (system.args.length >= 3) {
    url = system.args[2];
}

if (n <= 0) phantom.exit();

// Disable cookies
phantom.cookiesEnabled = false;

for (i = 0; i < n; i ++) {
    page = webpage.create();
    (function(pg, pgNum) {
        var clientId;

        pg.onConsoleMessage = function(msg) {
            if (msg === 'Game over') {
                // Game Over
                clientId = pg.evaluate(function() { return node.player.id; });
                console.log('Finished ' + pgNum + ' at ' +
                    (new Date).getTime() + ' with ID ' + clientId);

                countClosed ++;
                if (countClosed >= n) {
                    phantom.exit();
                }
            }
        }

        pg.open(url, function() {
            console.log('Opened ' + pgNum + ' at ' + (new Date).getTime());
        });
    })(page, i);
}


run_benchmark.py
#!/usr/bin/env python3

import re
import os
import sys
import json
import fileinput
import subprocess
import time
import csv
import platform
import datetime
import configparser
import argparse
import collections

try:
    import psutil
except ImportError:
    found_psutil = False
    print("Was not able to import psutil. Please install it via `pip3 install "
          "psutil`.\nThe benchmark will run, but it won't be able to extract "
          "CPU or memory metrics.\n", file=sys.stderr)
else:
    found_psutil = True


def get_cmd_args():
    # Define ArgumentParser and declare all needed command line arguments
    parser = argparse.ArgumentParser(description='Execute nodegame benchmark '
                                     'and write benchmark data to csv file.')

    parser.add_argument('-c', '--config', type=str, required=True,
                        help='Benchmark configuration file in INI format '
                        'containing variables that likely do not change '
                        'between benchmarks.')

    parser.add_argument('-n', '--num_conns', type=int, nargs='+',
                        help='Number of simultaneous connections to consider '
                        'for the benchmark, can be a list.')

    parser.add_argument('-r', '--reliable', action='store_true',
                        help='Boolean flag to turn on reliable messaging.')

    parser.add_argument('-nr', '--no_run', action='store_true',
                        help='Boolean flag to disable launching the game. '
                        'Will just process existing log files.')

    parser.add_argument('-t', '--timeouts', type=int, nargs='+',
                        help='Timeouts to consider for the benchmark when '
                        'reliable messaging is used, can be a list.')

    args = parser.parse_args()

    # Manually check dependency between command line arguments
    if not args.no_run:
        if not args.num_conns:
            print('Error: --num_conns needs to be specified when a benchmark '
                  'is run.', file=sys.stderr)
            sys.exit(1)

        if args.reliable and not args.timeouts:
            print('Error: --timeouts needs to be specified when reliable '
                  'messaging is activated.', file=sys.stderr)
            sys.exit(1)

    # Make sure we have a default value for args.timeouts. This is important
    # because we are iterating over it, even though the actual value does not
    # matter
    if not args.reliable:
        args.timeouts = [4000]

    return args


def expand_user_in_cfg(cfg):
    """ Iterate over all options in both the 'Directories' and 'Files' sections
    and expand the user variable"""
    for dir_option in cfg.options('Directories'):
        cfg.set('Directories', dir_option,
                os.path.expanduser(cfg.get('Directories', dir_option)))

    for file_option in cfg.options('Files'):
        cfg.set('Files', file_option,
                os.path.expanduser(cfg.get('Files', file_option)))


# Record the current Unix time in micro seconds.
# This is used to uniquely identify the benchmark.
BENCHMARK_TIME = int(time.time() * 10**6)


def get_benchmark_filename(folder, suffix, ext):
    """ Utility function to create benchmark filenames with timestamp included.
    """
    file_name = 'benchmark_{}_{}.{}'.format(BENCHMARK_TIME, suffix, ext)
    return os.path.join(folder, file_name)


def write_launcher_settings(settings_file, settings):
    with open(settings_file, 'w') as settings_fp:
        settings_str = ",\n".join(["    {}: {}".format(k, v)
                                  for (k, v) in settings])
        settings_fp.write("module.exports = {{\n{}\n}};\n"
                          .format(settings_str))


def write_timeout_to_cfg_files(cfg, reliable, timeout):
    """ Writes the retry timeout and the reliable boolean flag to the client
    and server var file. Note that even though timeout is written every time it
    only takes effect if reliable == True.
    """
    for mode in ['client', 'server']:
        var_section = '{} Variables'.format(mode.capitalize())
        var_file = '{}_var_file'.format(mode)

        re_reliable = re.compile(r'({0})\s*=\s*(true|false)'.format(
                                 cfg.get(var_section, 'rel_msg_var')))

        re_retry = re.compile(r'({0})\s*=\s*\d+'.format(
                              cfg.get(var_section, 'rel_retry_var')))

        # We iterate through the client variable file and modify it in-place.
        # In this case everything written to stdout will be redirected to the
        # file we opened, hence we need to print every line.
        for line in fileinput.input(cfg.get('Files', var_file), inplace=True):
            # Remove trailing whitespace
            line = line.rstrip()

            # If the current line matches the reliable regular expression, do
            # the appropriate substitution. We convert reliable to lower case,
            # because booleans are uppercase in python (e.g. True vs. true).
            if re_reliable.search(line):
                print(re_reliable.sub(r'\1 = ' + str(reliable).lower(), line))
            # Else if it matches the retry variable regular expression, do
            # another substitution.
            elif re_retry.search(line):
                print(re_retry.sub(r'\1 = ' + str(timeout), line))
            # Else print the original line.
            else:
                print(line)


def sizeof_fmt(num, suffix='B'):
    """ Utility function to convert byte amounts to human readable format.
    Taken from http://stackoverflow.com/a/1094933/2528077 """
    for unit in ['', 'Ki', 'Mi', 'Gi', 'Ti', 'Pi', 'Ei', 'Zi']:
        if abs(num) < 1024.0:
            return "%3.1f%s%s" % (num, unit, suffix)
        num /= 1024.0
    return "%.1f%s%s" % (num, 'Yi', suffix)


def time_fmt(seconds):
    """ Utilty function to convert duration to human readable format. Follows
    default format of the Unix `time` command. """
    return "{:.0f}m{:.3f}s".format(seconds // 60, seconds % 60)


def build_nodegame(cfg):
    """ Routine to build nodegame, saves the build log into a separate file.
    Warns if there was an error. """
    build_log = get_benchmark_filename(cfg.get('Directories', 'log_dir'),
                                       'build', 'log')

    print('Build Log:\n{}\n'.format(build_log))
    with open(build_log, 'a') as b_log:
        retcode = subprocess.call(['node', 'bin/make.js', 'build-client',
                                   '-a', '-o', 'nodegame-full'],
                                  cwd=cfg.get('Directories', 'server_dir'),
                                  stdout=b_log, stderr=b_log)

        if retcode:
            print("Warning: The nodegame build had a non-zero exit code.",
                  file=sys.stderr)


def run_launcher(cfg):
    """ Executes `node launcher.js` from the right cwd and logs stdout and
    stderr to the previously defined log folder.
    """

    stdout_log = get_benchmark_filename(cfg.get('Directories', 'log_dir'),
                                        'stdout', 'log')

    stderr_log = get_benchmark_filename(cfg.get('Directories', 'log_dir'),
                                        'stderr', 'log')

    print('Logging stdout and stderr:\n{}\n{}'.format(stdout_log, stderr_log))
    launcher_file = cfg.get('Files', 'launcher_file')
    if not os.path.exists(launcher_file):
        raise FileNotFoundError("$[Files] launcher_file = {} does not "
                                "exist.".format(launcher_file))

    launcher_cwd = cfg.get('Directories', 'launcher_cwd')
    if not os.path.exists(launcher_cwd):
        raise FileNotFoundError("[Directories] launcher_cwd = {} does not "
                                "exist.".format(launcher_cwd))

    with open(stdout_log, 'a') as f_out, open(stderr_log, 'a') as f_err:
        proc = subprocess.Popen(['node', cfg.get('Files', 'launcher_file'),
                                cfg.get('General Settings', 'game')],
                                cwd=cfg.get('Directories', 'launcher_cwd'),
                                stdout=f_out, stderr=f_err)

        return proc


def get_process_metrics(proc):
    """ Extracts CPU times, memory infos and connection infos about a given
    process started via Popen(). Also obtains the return code. """
    p = psutil.Process(proc.pid)
    max_cpu = [0, 0]
    max_mem = [0, 0]
    conns = []

    while proc.poll() is None:
        try:
            cpu = list(p.cpu_times())
            mem = list(p.memory_info())
            conns = p.connections('all')

            for child in p.children(recursive=True):
                c_cpu = list(child.cpu_times())
                c_mem = list(child.memory_info())

                cpu[0] += c_cpu[0]
                cpu[1] += c_cpu[1]

                mem[0] += c_mem[0]
                mem[1] += c_mem[1]

            if max_cpu[0] < cpu[0]:
                max_cpu = cpu

            if max_mem[0] < mem[0]:
                max_mem = mem

        except (psutil.AccessDenied, psutil.NoSuchProcess):
            pass
        time.sleep(1)
    retcode = proc.wait()

    return retcode, max_cpu, max_mem, conns


def run_test(cfg):
    """ Runs `npm test` from the correct cwd and returns the return code. """
    return subprocess.call(['npm', 'test'],
                           cwd=cfg.get('Directories', 'test_cwd'))


def parse_server_msg_file(msg_file, is_reliable):
    """ Parses the server message log file. Extract metrics about the total
    number of messages and the break down according to type. In addition
    computes the average delay of a message round-trip if reliable messaging is
    enabled. """

    # define a message counter and a timestamps dictionary for both client and
    # server
    msg_counter = collections.Counter()
    timestamps = {'client': {}, 'server': {}}

    # open the message file for reading
    with open(msg_file) as messages:
        for message in messages:
            # increment total message counter
            msg_counter['total'] += 1

            # parse the resulting json strings
            winston_msg = json.loads(message)
            game_msg = winston_msg['GameMsg']

            # increment corresponding target counter
            msg_counter[game_msg['target']] += 1

            # skip the rest if reliable messaging is not activated
            if not is_reliable:
                continue

            # extract message id
            msg_id = str(game_msg['id'])

            # parse JavaScript Date.prototype.toISOString() into a Python
            # datetime object
            created = datetime.datetime.strptime(game_msg['created'],
                                                 '%Y-%m-%dT%H:%M:%S.%fZ')

            timestamp = datetime.datetime.strptime(winston_msg['timestamp'],
                                                   '%Y-%m-%dT%H:%M:%S.%fZ')

            # initialize timestamps
            if msg_id not in timestamps['client']:
                timestamps['client'][msg_id] = [0, 0]
            if msg_id not in timestamps['server']:
                timestamps['server'][msg_id] = [0, 0]

            # different between ACK and normal messages for both client and
            # server
            if game_msg['target'] == 'ACK':
                if game_msg['to'] == 'SERVER':
                    timestamps['server'][game_msg['text']][1] = timestamp
                elif game_msg['from'] == 'ultimatum':
                    timestamps['client'][game_msg['text']][1] = timestamp

            else:
                if game_msg['to'] == 'SERVER':
                    timestamps['client'][msg_id][0] = created
                elif game_msg['from'] == 'ultimatum':
                    timestamps['server'][msg_id][0] = timestamp

    # simply return counter if no reliable messaging
    if not is_reliable:
        return msg_counter

    # compute timedeltas for both client and server
    client_server_times = [
        v[1] - v[0] for v in timestamps['client'].values() if v[0] and v[1]
    ]

    server_client_times = {
        v[1] - v[0] for v in timestamps['server'].values() if v[0] and v[1]
    }

    if len(client_server_times) == 0:
        print("Warning: Could not record time deltas for client -> server "
              "messages.", file=sys.stderr)
        avg_client_server_time = 0.0
    else:
        avg_client_server_time = sum(
            client_server_times, datetime.timedelta(0)
        ).total_seconds() / len(client_server_times)

    if len(server_client_times) == 0:
        print("Warning: Could not record time deltas for server -> client "
              "messages.", file=sys.stderr)
        avg_server_client_time = 0.0
    else:
        avg_server_client_time = sum(
            server_client_times, datetime.timedelta(0)
        ).total_seconds() / len(server_client_times)

    print("The average delay to deliver a message was {:.0f} milliseconds."
          .format(avg_server_client_time * 1000))
    return msg_counter, avg_client_server_time, avg_server_client_time


def main():
    args = get_cmd_args()

    with open(args.config) as cfg_fp:
        cfg = configparser.ConfigParser(interpolation=configparser.
                                        ExtendedInterpolation())
        # make the config options case sensitive
        cfg.optionxform = str
        cfg.read_file(cfg_fp)

    expand_user_in_cfg(cfg)

    # construct metrics.csv file name
    if args.no_run:
        csv_metrics_file = os.devnull
    else:
        csv_metrics_file = \
            get_benchmark_filename(cfg.get('Directories', 'csv_dir'),
                                   'metrics', 'csv')

    # construct messages.csv file name
    csv_msg_file = \
        get_benchmark_filename(cfg.get('Directories', 'csv_dir'),
                               'messages', 'csv')

    print('CSV files:\n{}\n{}\n'.format(csv_metrics_file, csv_msg_file))
    # this defines the metrics we want to record
    metrics_names = [
        "id", "machine", "num_conns", "is_reliable", "timeout",
        "benchmark_ret_code", "test_ret_code", "cpu_time_user",
        "cpu_time_system", "mem_info_rss", "mem_info_vms",
        "avg_client_time", "avg_server_time"
    ]

    # this defines the messages we want to record
    msg_names = [
        "id", "total",
        "ACK", "ALERT", "BYE", "DATA", "ERR", "GAMECOMMAND", "HI", "JOIN",
        "LANG", "LOG", "MCONNECT", "MDISCONNECT", "MLIST", "MRECONNECT",
        "PCONNECT", "PDISCONNECT", "PLAYER_UPDATE", "PLIST", "PRECONNECT",
        "REDIRECT", "SERVERCOMMAND", "SETUP", "STAGE", "STAGE_LEVEL",
        "TXT", "WARN"
    ]

    # open csv files for writing
    with open(csv_metrics_file, 'w') as csv_metrics, \
            open(csv_msg_file, 'w') as csv_msg:

        # define the respective csv writers and write the header rows
        metrics_writer = csv.DictWriter(csv_metrics, fieldnames=metrics_names)
        metrics_writer.writeheader()

        msg_writer = csv.DictWriter(csv_msg, fieldnames=msg_names)
        msg_writer.writeheader()
        msg_file = os.path.join(cfg.get("Directories", "msg_log_dir"),
                                cfg.get("Files", "server_msg_file"))

        if args.no_run:
            if args.reliable:
                msg_counter, avg_client_time, avg_server_time = \
                    parse_server_msg_file(msg_file, args.reliable)
            else:
                msg_counter = \
                    parse_server_msg_file(msg_file, args.reliable)

            # add 'id' field to the message counter
            msg_counter["id"] = BENCHMARK_TIME
            # we manually set not occurring counts to 0 to avoid empty
            # strings in the csv
            for msg_name in msg_names:
                if msg_name not in msg_counter:
                    msg_counter[msg_name] = 0

            # finally write the message statistics
            msg_writer.writerow(msg_counter)
            return

        # iterate over the number of connections
        for num_conns in args.num_conns:
            # set the current number of connections in the cfg object and write
            # it to the launcher settings file
            cfg.set('Launcher Settings', 'numPlayers', str(num_conns))
            write_launcher_settings(cfg.get('Files', 'launcher_settings_file'),
                                    cfg.items('Launcher Settings'))

            # iterate over the specified timeouts
            for timeout in args.timeouts:
                write_timeout_to_cfg_files(cfg, bool(args.reliable), timeout)
                # we try to delete the existing file, if this fails it means
                # it did not exist in the first place, so we can just continue
                try:
                    os.remove(msg_file)
                except OSError:
                    pass

                print("Building Client")
                build_nodegame(cfg)

                # run_timestamp serves as the current run id
                run_timestamp = int(time.time() * 10**6)

                # print information about the current run configuration to
                # standard output
                print("Running Benchmark")
                print("Number of Connections: {}, Reliable: {}, Timeout: {}"
                      .format(num_conns, bool(args.reliable), timeout))

                # start the launcher process
                launcher = run_launcher(cfg)

                # if psutil is installed record operating system utils
                if found_psutil:
                    ret_benchmark, cpu, mem, conns = \
                        get_process_metrics(launcher)
                # else just wait for termination of the run
                else:
                    ret_benchmark = launcher.wait()

                if ret_benchmark:
                    print("Warning: The current run had a non-zero exit code. "
                          "Please have a look at the log,\nthe benchmark id is"
                          " {}.".format(BENCHMARK_TIME), file=sys.stderr)

                time.sleep(1)
                ret_test = run_test(cfg)
                if ret_test:
                    print("Warning: The test run had a non-zero exit code.",
                          file=sys.stderr)

                # if reliable messaging is activated compute average response
                # time
                if args.reliable:
                    msg_counter, avg_client_time, avg_server_time = \
                        parse_server_msg_file(msg_file, args.reliable)
                else:
                    msg_counter = \
                        parse_server_msg_file(msg_file, args.reliable)

                # finally collect all benchmark metrics and write it to the csv
                benchmark_metrics = {
                    'id': run_timestamp,
                    'machine': platform.platform(),
                    'num_conns': num_conns,
                    'is_reliable': bool(args.reliable),
                    'timeout': timeout if args.reliable else 'NA',
                    'benchmark_ret_code': ret_benchmark,
                    'test_ret_code': ret_test,
                    'cpu_time_user':
                        time_fmt(cpu[0]) if found_psutil else 'NA',
                    'cpu_time_system':
                        time_fmt(cpu[1]) if found_psutil else 'NA',
                    'mem_info_rss':
                        sizeof_fmt(mem[0]) if found_psutil else 'NA',
                    'mem_info_vms':
                        sizeof_fmt(mem[1]) if found_psutil else 'NA',
                    'avg_client_time':
                        time_fmt(avg_client_time) if args.reliable else 'NA',
                    'avg_server_time':
                        time_fmt(avg_server_time) if args.reliable else 'NA'
                }

                metrics_writer.writerow(benchmark_metrics)

                # add 'id' field to the message counter
                msg_counter["id"] = run_timestamp
                # we manually set not occurring counts to 0 to avoid empty
                # strings in the csv
                for msg_name in msg_names:
                    if msg_name not in msg_counter:
                        msg_counter[msg_name] = 0

                # finally write the message statistics
                msg_writer.writerow(msg_counter)

if __name__ == '__main__':
    try:
        main()
        error = False
    except configparser.ParsingError as err:
        error = err
    except configparser.NoSectionError as err:
        print("Error: The config file has a missing section.", file=sys.stderr)
        error = err
    except configparser.NoOptionError as err:
        print("Error: The config file has a missing option.", file=sys.stderr)
        error = err
    except (PermissionError, FileNotFoundError) as err:
        print("Error: The config file has an invalid value.", file=sys.stderr)
        error = err

    if error:
        print(error, file=sys.stderr)
        sys.exit(1)
